### PR TITLE
Synergies (upvotes) and edge now exist on back end

### PR DESCRIPTION
**Pull request by jesterswilde**
-Created a new table for relations between 2 cards called cardPairs
-Card pairs store the resulting data from join tables (currently synergies and edges)

-Edges are automatically generated when the user goes from one card to another.
-Synergies are upvotes on relationships between one card and another

-cardPairs are uni-directional. Meaning there are 2 entires for each card pair. One from cardA to cardB and one form cardB to cardA. In general the reason for this is a higher level of granularity on our understanding of user behvavior. I imagine a good amount of this will be abstracted away on the front end.

### CODE REVIEW

File: app/database/associations.js [hunk 1]
```diff
@@ -15,4 +17,18 @@ module.exports = ({card, user, comment, rating, type, subtype})=>{
...
+	cardPair.belongsTo(card, {foreignKey: 'cardID'}); 
+	cardPair.belongsTo(card, {foreignKey: 'otherCardID'});
+
+	cardPair.hasMany(cardEdge, {foreignKey: 'pairID'});
+	cardEdge.belongsTo(cardPair, {foreignKey: 'pairID'});
+	user.hasMany(cardEdge, {foreignKey: 'userID'}); 
+
+	user.hasMany(upvote, {foreignKey: 'userID'}); 
+	upvote.belongsTo(user, {foreignKey: 'userID'});
+	//upvotes can use many things as their parent key, such synergies and comments
```
Comment (jesterswilde) [inline app/database/associations.js]: By not forcing a foreign key here we may lose some cascading effect. I'm not sure how we'd keep this table generic and have a foreign key. We could have  a seperate table for each kind of upvote, but i do like this general behavior.

File: app/database/cardPair.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        autoIncrement: true
+    },
+    cardID:{
+      type: Sequelize.INTEGER,
+      unique: 'pair'
+    },
+    otherCardID:{
+      type: Sequelize.INTEGER,
+      unique: 'pair'
+    },
```
Comment (jesterswilde) [inline app/database/cardPair.js]: By feeding a string into the 'unique' property a  composite constraint is created by all  fields sharing the same 'unique' string

File: app/database/enums/edgeEnum.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+module.exports = [
+	'TO',
+	'FROM',
+	'RANDOM_TO',
+	'RANDOM_FROM',
```
Comment (jesterswilde) [inline app/database/enums/edgeEnum.js]: Random is added just to keep us thinking about that feature.

File: app/database/enums/voteTypes.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+module.exports = [
+	'SYNERGY',
+	'USR_CMMNT', //COMMENT is substring'd out of sequelize
```
Comment (jesterswilde) [inline app/database/enums/voteTypes.js]: 'COMMENT' is not allowed in sequelize Enums.

File: app/database/upvote.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+const votingOnEnum = require('./enums/voteTypes'); 
+const types = require('./enums/types'); 
+
+module.exports = (sequelize, Sequelize)=> {
+  return sequelize.define('upvote', {
+    subjectID:{
+      type: Sequelize.INTEGER,
+      primary: true,
+    },
```
Comment (jesterswilde) [inline app/database/upvote.js]: I do call it entityID elsewhere. Maybe I should make it all the same?

Comment (jonathanhawleypeters) [reply app/database/upvote.js]: Yes definitely

File: app/graphql/resolvers/cardEdge.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
+const { verifyToken } = require('../../auth/auth');
+const { cardPair, cardEdge, sequelize } = require('../../database/db');
+
+const createEdge = async (cardID, otherCardID, direction, userID = null)=>{
+	let pair = await cardPair.findOrCreate(
```
Comment (jonathanhawleypeters) [inline app/graphql/resolvers/cardEdge.js]: Why not:
``` Javascript
[ pair ] = await cardPair...
```

File: app/graphql/resolvers/upvote.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    return result; 
+  }catch(error){
+    console.error(error); 
+    return { error };
+  }
+};
+
+const createSynergyUpvoteMutation = async (obj, {cardID, otherCardID, ...unusedProps}, context, info)=>{
+  try{
+    const pair = await cardPair.findOrCreate({
```
Comment (jonathanhawleypeters) [inline app/graphql/resolvers/upvote.js]: Destructure yo pair foo'

**Pull request merged**